dev/com.ibm.ws.logging.core/src/com/ibm/ws/collector/manager/buffer/BufferManagerImpl.java

```diff
@@ -16,278 +16,278 @@
-                lostEventsForWarning = 0;
-            }
-            timeElapsed = currentTime - lastReportTimeForTrace;
-            if (timeElapsed >= intervalForTrace) {
-                if (TraceComponent.isAnyTracingEnabled() && tc.isEventEnabled()) {
-                    long timeElapsedInSecs = TimeUnit.MILLISECONDS.toSeconds(timeElapsed);
-                    Tr.event(tc,
-                            "Handler [{0}] has lost {1} events from source [{2}] in the last {3} second(s), and has lost {4} events from the source since the handler started.",
-                            handlerId, lostEventsForTrace, sourceId, timeElapsedInSecs, totalLostEvents);
-                }
-                lastReportTimeForTrace = currentTime;
-                lostEventsForTrace = 0;
-            }
-        }
-
-    }
+	private static final TraceComponent tc = Tr.register(BufferManagerImpl.class);
+	private Buffer<Object> ringBuffer;
+	private Set<SynchronousHandler> synchronousHandlerSet = new HashSet<SynchronousHandler>();
+
+	private final int capacity;
+
+	private final String sourceId;
+	/* Map to keep track of the next event for a handler */
+	private final ConcurrentHashMap<String, HandlerStats> handlerEventMap = new ConcurrentHashMap<String, HandlerStats>();
+
+	protected Queue<Object> earlyMessageQueue;
+
+	private static final int EARLY_MESSAGE_QUEUE_SIZE = 400;
+
+	public BufferManagerImpl(int capacity, String sourceId) {
+		super();
+
+		BufferManagerEMQHelper.addBufferManagerList(this);
+		ringBuffer = null;
+		this.sourceId = sourceId;
+		this.capacity = capacity;
+		if (!BufferManagerEMQHelper.getEMQRemovedFlag()) {
+			earlyMessageQueue = new SimpleRotatingSoftQueue<Object>(new Object[EARLY_MESSAGE_QUEUE_SIZE]);
+			//Check again just in case
+			if (BufferManagerEMQHelper.getEMQRemovedFlag()){
+				removeEMQ();
+			}
+		}
+	}
+
+	public BufferManagerImpl(int capacity, String sourceId, boolean isEMQ) {
+		super();
+		this.sourceId = sourceId;
+		this.capacity = capacity;
+		if (!isEMQ) {
+			earlyMessageQueue = null; // don't need earlyMessageQueue
+			ringBuffer = new Buffer<Object>(capacity);
+		}
+	}
+
+	@Override
+	public void add(Object event) {
+		if (event == null)
+			throw new NullPointerException();
+
+		
+		if (ringBuffer != null) {
+			/* ringBuffer was first checked to be not null, but circumstances
+			 * may occur where ringBuffer can be null after the check.
+			 * 
+			 * The following method "addEventToRingBuffer" is used to avoid a NPE
+			 */
+			addEventToRingBuffer(event);
+		}
+
+		if (earlyMessageQueue != null) {
+			/*
+			 * earlyMessageQueue was first checked to be not null but this "gap"
+			 * here may have allowed the earlyMessageQueue to be set to null
+			 * from removeEMQ()
+			 */
+			synchronized (this) {
+				// Must check again
+				if (earlyMessageQueue != null) {
+					earlyMessageQueue.add(event);
+				}
+			}
+		}
+
+		for (SynchronousHandler synchronousHandler : synchronousHandlerSet) {
```

Comment:
if synchronousHandler is added on another thread while this thread is between lines 94-97, that synchronousHandler will get this event twice.